### PR TITLE
OI-159 Implement: "user_mention" Message template, Rules action - "Cr…

### PR DESCRIPTION
**Pull request by nvelychenko**
…eate message after mentioning user". Add appropriate message_template texts for the message displayed itslef, email subject and body. Add additional conditions for "activity feed" views. Rework the MessageAction.

### CODE REVIEW

**Comment by db-Matroskeen** [comment]
@nvelychenko,
I'm just curious to know how it'll work in the following cases:
1) Mention user in the existing comment (via "Edit" action) - will the event trigger;
2) Add a comment with the mentioned user, and then remove mention (via "Edit" action);
3) Do users have the ability to remove comments? If so, what happens if they add a comment with mentioned user and then remove this comment after some time?

4) Also, what happens when I'm looking at someone else's page? Will I see ".. has mentioned **you** on ..." in that case?

**Comment by nvelychenko** [comment]
> @nvelychenko,
> I'm just curious to know how it'll work in the following cases:
> 
> 1. Mention user in the existing comment (via "Edit" action) - will the event trigger;
> 2. Add a comment with the mentioned user, and then remove mention (via "Edit" action);
> 3. Do users have the ability to remove comments? If so, what happens if they add a comment with mentioned user and then remove this comment after some time?
> 4. Also, what happens when I'm looking at someone else's page? Will I see ".. has mentioned **you** on ..." in that case?

1. Nope, there is another event during entity updating.
2. Nothing will happen. See 1 point.
3. We have logic to remove all orphaned message entities, when a comment, node, or user entity is deleted.
4. Fixed to show **you** for the current user page, and **mentioned user** for other profiles.

File: themes/openideal_theme/openideal_theme.theme [hunk 1]
```diff
@@ -185,3 +185,18 @@ function openideal_theme_preprocess_image(&$variables) {
...
+ * Implements hook_preprocess_HOOK().
+ */
+function openideal_theme_preprocess_message__user_mention(&$variables) {
+  // In case the user is not viewing its page,
+  // its necessary to change the partial template.
+  if (\Drupal::routeMatch()->getRouteName() == 'entity.user.canonical' &&
+    \Drupal::currentUser()->id() !== \Drupal::routeMatch()->getParameter('user')->id()) {
+    $elements = &$variables['elements'];
+    $variables['#cache']['contexts'][] = 'user.is_user_profile_page';
+    $partials = $elements['#message']->getText(\Drupal::languageManager()->getCurrentLanguage()->getId());
```
Comment (db-Matroskeen) [inline themes/openideal_theme/openideal_theme.theme]: `getText()` accepts one more argument - delta. Can we pass "3" to get correct text line?

File: themes/openideal_theme/openideal_theme.theme [hunk 2]
```diff
@@ -185,3 +185,18 @@ function openideal_theme_preprocess_image(&$variables) {
...
+/**
+ * Implements hook_preprocess_HOOK().
+ */
+function openideal_theme_preprocess_message__user_mention(&$variables) {
+  // In case the user is not viewing its page,
+  // its necessary to change the partial template.
+  if (\Drupal::routeMatch()->getRouteName() == 'entity.user.canonical' &&
+    \Drupal::currentUser()->id() !== \Drupal::routeMatch()->getParameter('user')->id()) {
+    $elements = &$variables['elements'];
+    $variables['#cache']['contexts'][] = 'user.is_user_profile_page';
```
Comment (db-Matroskeen) [inline themes/openideal_theme/openideal_theme.theme:198]: We have to merge cache metadata (not overwrite). See `Drupal\Core\Cache\Cache.php` for helper methods.

File: themes/openideal_theme/openideal_theme.theme [hunk 3]
```diff
@@ -185,3 +185,18 @@ function openideal_theme_preprocess_image(&$variables) {
...
+
+/**
+ * Implements hook_preprocess_HOOK().
+ */
+function openideal_theme_preprocess_message__user_mention(&$variables) {
+  // In case the user is not viewing its page,
+  // its necessary to change the partial template.
+  if (\Drupal::routeMatch()->getRouteName() == 'entity.user.canonical' &&
+    \Drupal::currentUser()->id() !== \Drupal::routeMatch()->getParameter('user')->id()) {
+    $elements = &$variables['elements'];
```
Comment (db-Matroskeen) [inline themes/openideal_theme/openideal_theme.theme]: It seems ampersand is not necessary here.

**Review by db-Matroskeen — CHANGES_REQUESTED** [review]

**Pull request merged**